deepchecks/vision/dataset.py

```diff
@@ -0,0 +1,140 @@
+from collections import Counter
+from typing import Callable
+
+from torch.utils.data import DataLoader
+from torch import cat
+import logging
+
+from deepchecks.core.errors import DeepchecksValueError
+
+logger = logging.getLogger('deepchecks')
+
+__all__ = ['TaskType', 'VisionDataset']
+
+
+class TaskType(Enum):
+    """Enum containing supported task types."""
+
+    CLASSIFICATION = 'classification'
+    OBJECT_DETECTION = 'object_detection'
+    SEMANTIC_SEGMENTATION = 'semantic_segmentation'
+
+
+class VisionDataset:
+
+    _data: DataLoader = None
+
+    def __init__(self, data_loader: DataLoader, num_classes: int = None, label_type: str = None,
+                 label_transformer: Callable = None):
+        self._data = data_loader
+
+        if label_transformer is None:
+            self.label_transformer = lambda x: x
+        else:
+            self.label_transformer = label_transformer
+
+        if label_type is not None:
+            self.label_type = label_type
+        else:
+            self.label_type = self.infer_label_type()
+
+        self._num_classes = num_classes  # if not initialized, then initialized later in get_num_classes()
+        self._samples_per_class = None
+
+    def get_num_classes(self):
+        if self._num_classes is None:
+            samples_per_class = self.get_samples_per_class()
+            num_classes = len(samples_per_class.keys())
+            self._num_classes = num_classes
+        return self._num_classes
+
+    def get_samples_per_class(self):
+        if self._samples_per_class is None:
+            if self.label_type == TaskType.CLASSIFICATION.value:
+                counter = Counter()
+                for _ in range(len(self._data)):
+                    counter.update(self.label_transformer(next(iter(self._data))[1]))
+                self._samples_per_class = counter
+            elif self.label_type == TaskType.OBJECT_DETECTION.value:
+                # Assume next(iter(self._data))[1] is a list (per sample) of numpy arrays (rows are bboxes) with the
+                # first column in the array representing class
+                counter = Counter()
+                for i in range(len(self._data)):
+                    list_of_arrays = self.label_transformer(next(iter(self._data))[1])
+                    class_list = sum([arr.reshape((-1, 5))[:, 0].tolist() for arr in list_of_arrays], [])
+                    counter.update(class_list)
+                self._samples_per_class = counter
+            else:
+                raise NotImplementedError(
+                    'Not implemented yet for tasks other than classification and object detection'
+                )
+        return copy(self._samples_per_class)
+
+    def extract_label(self):
+        y = []
+        for _ in range(len(self._data)):
+            y.append(self.label_transformer(next(iter(self._data))[1]))
+        return cat(y, 0)
+
+    def infer_label_type(self):
+        label_shape = self.get_label_shape()
```

Comment:
feel that given the label_transformer, we should get rid of this. It's the user's responsibility from now on to transform the label to a fitting format. 

It is our responsibility to check that the resulting output matches our requirement, so we should add validation for that instead of this.